src/crypto/mod.rs

```diff
@@ -81,17 +80,33 @@ impl EcKeyComponents {
     }
 }
 
-pub trait Crypto: Sized {
-    type RemotePublicKey: RemotePublicKey;
-    type LocalKeyPair: LocalKeyPair;
-    fn generate_ephemeral_keypair() -> Result<Self::LocalKeyPair>;
+pub trait Cryptographer: Send + Sync + 'static {
+    /// Generate a random ephemeral local key pair.
+    fn generate_ephemeral_keypair(&self) -> Result<Box<dyn LocalKeyPair>>;
+    /// Import a local keypair from its raw components.
+    fn import_key_pair(&self, components: &EcKeyComponents) -> Result<Box<dyn LocalKeyPair>>;
+    /// Import the public key component in the binary uncompressed point representation.
+    fn import_public_key(&self, raw: &[u8]) -> Result<Box<dyn RemotePublicKey>>;
     fn compute_ecdh_secret(
-        remote: &Self::RemotePublicKey,
-        local: &Self::LocalKeyPair,
+        &self,
+        remote: &RemotePublicKey,
+        local: &LocalKeyPair,
     ) -> Result<Vec<u8>>;
-    fn hkdf_sha256(salt: &[u8], secret: &[u8], info: &[u8], len: usize) -> Result<Vec<u8>>;
+    fn hkdf_sha256(&self, salt: &[u8], secret: &[u8], info: &[u8], len: usize) -> Result<Vec<u8>>;
     /// Should return [ciphertext, auth_tag].
-    fn aes_gcm_128_encrypt(key: &[u8], iv: &[u8], data: &[u8], tag_len: usize) -> Result<Vec<u8>>;
-    fn aes_gcm_128_decrypt(key: &[u8], iv: &[u8], data: &[u8], tag: &[u8]) -> Result<Vec<u8>>;
-    fn random(dest: &mut [u8]) -> Result<()>;
+    fn aes_gcm_128_encrypt(
+        &self,
+        key: &[u8],
+        iv: &[u8],
+        data: &[u8],
+        tag_len: usize,
```

Comment:
Since you're churning the API anyway, should we consider dropping this `tag_len` argument?  IIRC there is only ever one value that is valid to pass in here, so we should just provide it automatically.